### PR TITLE
relocation elision

### PR DESCRIPTION
**Pull request by ecatmur**
Allow elision of relocation

### CODE REVIEW

File: relocation.bs [hunk 1]
```diff
@@ -733,6 +733,23 @@ void do_something_03(std::mutex& m)
 }
 ```
 
+### Relocation elision ### {#reloc-elision}
+
+The above elision is a specific case of a more general rule, which is that
+relocation (whether performed by relocation or move constructor)
```
Comment (SebastienBini) [inline relocation.bs:739]: why not by copy?

File: relocation.bs [hunk 2]
```diff
@@ -733,6 +733,23 @@ void do_something_03(std::mutex& m)
 }
 ```
 
+### Relocation elision ### {#reloc-elision}
```
Comment (SebastienBini) [inline relocation.bs:736]: This section is a nice addition, but I would move it before "Discarded reloc expressions".

File: relocation.bs (hunk 1 above)
Comment (ecatmur) [reply relocation.bs:739]: Indeed. See #15.

File: relocation.bs [hunk 3]
```diff
@@ -698,17 +714,17 @@ constructed from the source object. The temporary is then destructed at the end
...
+elide the creation
+of the temporary object, 
 effectively only calling the destructor of the source
 object at the end of the expression evaluation. If this optimization is done, 
 the temporary returned by `reloc` is not materialized.
 
 This means that `reloc obj;` has the following behavior:
 
 - if `obj` is not an *unowned parameter* then:
-    - If `obj` has a relocation or move constructor,
```
Comment (ecatmur) [inline relocation.bs:710]: I didn't understand this part. Is this related to #15 ?

Comment (SebastienBini) [reply relocation.bs:710]: Yes, the goal was to tackle this issue.
